walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,132 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<String> attendees = request.getAttendees();
+    Collection<String> optionalAttendees = request.getOptionalAttendees();
+    long duration = request.getDuration();
+
+    if (attendees.isEmpty() && optionalAttendees.isEmpty()){
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }else 
+    if (duration > TimeRange.WHOLE_DAY.duration()){
+      return Arrays.asList();
+    }
+
+    List<TimeRange> takenTime = new ArrayList<TimeRange>();
+    List<TimeRange> optionalBusy = new ArrayList<TimeRange>();
+    setTakenTimes(events, takenTime, optionalBusy, attendees, optionalAttendees);
```

Comment:
Based on the method name "query", why we need to "setTakenTimes" and "setFreeTimes". Isn't that something be done in a method called "book"?

Reply:
The method query is to determine what time of day the listed attendees are free. Thus, I set the takenTimes to know when they're unavailable and set the freeTimes to know when they are available.